charger/openwb-native_linux.go

```diff
@@ -0,0 +1,191 @@
+
+	evse, err := NewEvseDIN(ctx, uri, device, comset, baudrate, proto, slaveID)
+	if err != nil {
+		return nil, err
+	}
+
+	chargeState, err := evse.Status()
+	if err != nil {
+		return nil, err
+	}
+
+	wb := &OpenWbNative{evse, log, nil, "", cpWait, chargePoint, chargeState}
+
+	var (
+		phases1p3p func(int) error
+		identify   func() (string, error)
+	)
+
+	// configure special external hardware features
+	if hasPhases1p3p {
+		phases1p3p = wb.phases1p3p
+	}
+
+	if rfIdVidPid != "" {
+		rfIdChannel, _, err := native.NewRFIDHandler(rfIdVidPid, ctx, log)
+		if err != nil {
+			return nil, err
+		}
+		wb.rfIdChannel = rfIdChannel
+
+		identify = wb.identify
+	}
+
+	return decorateOpenWbNative(wb, phases1p3p, identify), nil
+}
+
+// Status implements the api.Charger interface
+func (wb *OpenWbNative) Status() (api.ChargeStatus, error) {
+	res, err := wb.Charger.Status()
+	if wb.chargeState != api.StatusA && res == api.StatusA {
+		// Status changed from connected/charging to not connected, discard rfid
+		wb.rfId = ""
+	}
+	return res, err
+}
+
+// phases1p3p implements the api.PhaseSwitcher interface
+func (wb *OpenWbNative) phases1p3p(phases int) error {
+	return wb.gpioExecute(func() error {
+		return wb.gpioSwitchPhases(phases)
+	})
+}
+
+var _ api.Resurrector = (*OpenWbNative)(nil)
+
+// WakeUp implements the api.Resurrector interface
+func (wb *OpenWbNative) WakeUp() error {
+	return wb.gpioExecute(wb.gpioWakeup)
+}
+
+func (wb *OpenWbNative) gpioExecute(worker func() error) error {
+	if err := wb.Enable(false); err != nil {
+		return err
+	}
+
+	if err := rpio.Open(); err != nil {
+		return err
+	}
+	defer rpio.Close()
+
+	if err := worker(); err != nil {
+		return err
+	}
+
+	if err := wb.Enable(true); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Worker function to toggle the GPIOs to switch the phases
+func (wb *OpenWbNative) gpioSwitchPhases(phases int) error {
+	pinGpioCP := rpio.Pin(native.ChargePoints[wb.chargePoint].PIN_CP)
+	pinGpioPhases := rpio.Pin(native.ChargePoints[wb.chargePoint].PIN_3P)
+	if phases == 1 {
+		pinGpioPhases = rpio.Pin(native.ChargePoints[wb.chargePoint].PIN_1P)
+	}
+	pinGpioCP.Output()
+	pinGpioPhases.Output()
+
+	time.Sleep(time.Second)
+	pinGpioCP.High() // enable phases switch relay (NO), disconnect CP
+
+	time.Sleep(time.Second)
+	pinGpioPhases.High() // move latching relay to desired position
+
+	time.Sleep(wb.cpWait / 2)
+	pinGpioPhases.Low() // lock latching relay
+
+	time.Sleep(wb.cpWait / 2)
+	pinGpioCP.Low() // disable phase switching, reconnect CP
+
+	time.Sleep(time.Second)
+	return nil
+}
+
+// Worker function to toggle the GPIOs for the CP signal
+func (wb *OpenWbNative) gpioWakeup() error {
+	pinGpioCP := rpio.Pin(native.ChargePoints[wb.chargePoint].PIN_CP)
+	pinGpioCP.Output()
+
+	pinGpioCP.High()
+	time.Sleep(wb.cpWait)
+	pinGpioCP.Low()
+	return nil
+}
+
+// Identify implements the api.Identifier interface
+func (wb *OpenWbNative) identify() (string, error) {
+	for {
```

Comment:
Use mutex instead of channel plus synchronous read? You code here is actually blocking and probably not doing what you think it does.

Reply:
Updated